### PR TITLE
[IMPROVED] Memory allocations and pressure in the raft layer.

### PR DESCRIPTION
**Pull request by derekcollison**
Improved some hot spots on the memory subsystem with sync.Pools.

Signed-off-by: Derek Collison <derek@nats.io>

### CODE REVIEW

File: server/filestore.go [hunk 1]
```diff
@@ -4483,31 +4483,50 @@ func (mb *msgBlock) msgFromBuf(buf []byte, sm *StoreMsg, hh hash.Hash64) (*Store
 	return sm, nil
 }
 
+// Used to intern strings for subjects.
+// Based on idea from https://github.com/josharian/intern/blob/master/intern.go
```
Comment (neilalexander) [inline server/filestore.go:4487]: Well this is a pretty clever approach!

File: server/raft.go [hunk 2]
```diff
@@ -1989,16 +2055,25 @@ func (ar *appendEntryResponse) encode(b []byte) []byte {
 	return buf[:appendEntryResponseLen]
 }
 
+// Track all peers we may have ever seen to use an string interns for appendEntryResponse decoding.
+var peers sync.Map
```
Comment (neilalexander) [inline server/raft.go:2059]: It looks like nothing ever gets removed from this `sync.Map`?

**Review by neilalexander — APPROVED** [review]
LGTM overall

File: server/raft.go (hunk 2 above)
Comment (derekcollison) [reply server/raft.go:2059]: For peers, yes, even on a large system like NGS with peer replacements would be like 40 entries.

Comment (derekcollison) [reply server/raft.go:2059]: We could add a remove in there for peer remove entry..

Comment (derekcollison) [reply server/raft.go:2065]: Added.

**Pull request merged**